omnilib/prototypes/functions/functions-misc.lua

```diff
@@ -823,6 +835,137 @@ function omni.lib.find_entity_prototype(item)
 	return nil
 end
 
+function omni.lib.find_recipe(item)
+	if type(item)=="table" then return item elseif type(item)~="string" then return nil end
+	for _, p in pairs(data.raw.recipe) do
+		if standardized_recipes[p.name] == nil then omni.marathon.standardise(p) end
+		for _,r in pairs(p.normal.results) do
+			if r.name == item then return p end
+		end
+	end
+	--log("Could not find "..item.."'s recipe prototype, check it's type.")
+	return nil
+end
+
+-----------------------------------------------------------------------------
+-- ICON FUNCTIONS --
+-----------------------------------------------------------------------------
+
+--If our prototype doesn't have an icon, we need to find one that does
+local find_icon = function(item)
+	for _, p in pairs({"item","mining-tool","gun","ammo","armor","repair-tool","capsule","module","tool","rail-planner","item-with-entity-data","fluid","technology"}) do
+		if data.raw[p][item] then
+			if data.raw[p][item].icons or data.raw[p][item].icon then
+				return data.raw[p][item]
+			end
+		end
+	end
+end
+
+--really dig deep for the icon set
+local function find_result_icon(raw_item)
+	if raw_item then
+		if type(raw_item) ~= "table" then
+			raw_item = find_icon(raw_item) --Find a matching prototype if possible
+			return find_result_icon(raw_item)
+		elseif raw_item.icons then
+			local icons = table.deepcopy(raw_item.icons)
+			for i, icon in pairs(icons) do-- Apply inherited attributes as explicit for each layer
+				icon.icon_size = icon.icon_size or raw_item.icon_size
+			end
+			return icons
+		elseif raw_item.icon then
+			return {{
+				icon = raw_item.icon,
+				icon_size = raw_item.icon_size,
+				icon_mipmaps = raw_item.icon_mipmaps-- or nil
+			}}
+		else
+			local result = (-- recipe.result, first entry in recipe.results or either of the previous two within normal and expensive recipe blocks
+				(raw_item.result) or
+				(raw_item.results and raw_item.results[1].name) or
+				(raw_item.normal and (raw_item.normal.result or raw_item.normal.results[1].name)) or
+				(raw_item.expensive and (raw_item.expensive.result or raw_item.expensive.results[1].name)) 
+			)
+			return find_result_icon(result)
+		end
+	else
+		return {{
+			icon = "__core__/graphics/too-far.png",--ERROR
+			icon_size = 32,
+			icon_mipmaps = 1
+		}}
+	end
+end
+
+omni.lib.find_result_icon = find_result_icon
```

Comment:
Why is the function itself local?

Reply:
b116e4a, I had specified it wrong as local x = function would break tail calls, but it was an oversight on my part.

Reply:
Whoops, actually aa8f867